### ISSUE TITLE
Bug: Issue with dynamic prompts + hi-res fix + batch count > 1 (in A1111 / unclear beyond that)

### ISSUE THREAD

**Issue by shiverczar** [adieyal/sd-dynamic-prompts#805]
When using the A1111 UI to run stable diffusion, if you utilize dynamic prompts and hi-res fix with a batch count greater than 1, the dynamic prompts will be evaluated and saved/re-used in an unexpected manner.

I didn't test with batch SIZES greater than 1, unclear if it would impact the operation.

The first image will have its prompt evaluated and replaced properly, and utilized for both the initial generation and the hi-res fix. This is expected and desired behavior.

Images 2+ will have their prompt evaluated and replaced correctly, and utilized in the initial generation (good). The hi-res fix for images 2+ will all utilize the prompt from image 1 (bad).

Example, batch count 3 - The dynamic prompt will be """""randomly""""" evaluated in the example ;)
Prompt:
A {red|green|blue} flower grows in a pot

Image 1 base prompt: A red flower grows in a pot / Image 1 hi-res prompt: A red flower grows in a pot
Image 2 base prompt: A green flower grows in a pot / Image 2 hi-res prompt: A red flower grows in a pot
Image 3 base prompt: A blue flower grows in a pot / Image 3 hi-res prompt: A red flower grows in a pot

Having simple colors mix like this may not be a big deal - and with an image of a flower it may even come out well, but in larger and more complicated dynamic prompts it can be pretty problematic.

---

I will be the first to admit I am utterly clueless on how all of this works, but I believe replication of the issue should be straightforward. Just generate with some dynamic prompting, hi-res fix enabled, a batch count > 1, and toss the *second or later* image into PNG info -> send to txt2img to confirm

---

Workaround: Don't generate more than 1 batch at a time -- hi-res fix will use the base prompt for each image as expected. Generate Forever will not have the same problem, as it's the same as just generating 1 image at a time and then telling the UI to do it again after it's finished.

---
Generated a batch count 3 of my example. More-or-less default settings.

It evaluated the color for each as: blue/red/red.

Threw the third image into PNG Info:
![brave_Ua2t2EWDm1](https://github.com/user-attachments/assets/f10b693d-c15f-430b-8443-064e0d80a324)

As you can see, it saved 'a red flower in a pot' as the prompt, saved the template, and saved 'a blue flower in a pot' which was incorrectly added to it as the hires prompt, saved from the first batch of the 3 batch count.

It didn't seem to affect the image much in this case, it's clearly a red flower and there's no real influence of blue - maybe the pot in front? It was a very obviously green pot in the pre-hires-fix image (I have my A1111 set to save a copy of the image as-is before going into hires), it's now a sort of blue-green/gray.

**Comment by souljadogg** [comment]
Same issue, image1 - ok, starting from the second one - in hi-res fix all the prompts are from the image1, any solutions?

**Comment by Panchovix** [comment]
Issue still persist as nov 2024

**Comment by souljadogg** [comment]
I found a solution/compromise, just using other wildcards: [stable-diffusion-webui-wildcards](https://github.com/AUTOMATIC1111/stable-diffusion-webui-wildcards) 

cons: "|" does not work, using text files

**Comment by Panchovix** [comment]
The compromise is like not being able to use dynamic prompts on wildcards? like
`{__wildcard1__|__wildcard2__}`

**Comment by souljadogg** [comment]
Yep, but it works fine with 69kb files (tested), biggest I have.

**Comment by Panchovix** [comment]
But do you use it alongside with dynamic promtps extension? Don't they collide each other?
Also I guess if it works on that implementation, that somehow could be adapter to fix dynamic prompts (a big if though)

**Comment by souljadogg** [comment]
Nope, I disabled dynamic promtps.

**Comment by Iwannaknow800** [comment]
Just tried updating and I'm still having the same issue. When generating with batch count > 1 it uses the prompt from the first image as the hires prompt for every other image you generate after the first. This has been messing my generations up for a while without me realising. 

I really hope this can be fixed soon, it basically ruins batch count with hires as it is now.

**Comment by Cranbert** [comment]
Still an issue. Renders the extension kinda useless.
The wildcards extension requires you to make a txt file for every tiny little thing.
Generate forever has been broken on forge since summer.
This is so frustrating.

**Comment by Cranbert** [comment]
Okay, this is kinda bizarre. I was having the same issue, with Forge, where multiple batches would all get the same variables as the first in the hires fix field, screwing up all the generations other than the first.
I had tried updating the extension many times, and that never fixed it. Finally, I tried completely deleting and reinstalling the extension, and that actually did fix it... _mostly_.
Now, my generations still all show the same incorrect hrfix prompt in the metadata, but they actually generate with the correct prompt. So the images are no longer messed up, just the metadata. Whatever, not a big deal.
It seems the issue was forge somehow botching the update.

**Comment by Shoot2Skoot** [comment]
I found a fix via ChatGPT, but am not a coder, so please be careful if you choose to use this.

In [Forge]\webui\extensions\sd-dynamic-prompts\sd_dynamic_prompts\dynamic_prompting.py, you'll find this function:

`def _get_hr_fix_prompts(
    prompts: list[str],
    original_hr_prompt: str,
    original_prompt: str,
) -> list[str]:
    if original_prompt == original_hr_prompt:
        return list(prompts)
    return repeat_iterable_to_length([original_hr_prompt], len(prompts))
`

From what I understand this is checking for a cached hr_prompt and if one exists, it uses that rather than the current prompt. (or something similar)

Replacing this function with the following corrected the issue for me:
`
def _get_hr_fix_prompts(
    prompts: list[str],
) -> list[str]:
    return list(prompts)
`

In my case I left the original function as a comment so I can refer back to it.

`
def _get_hr_fix_prompts(
    prompts: list[str],
) -> list[str]:
    return list(prompts)

""" Original Function

def _get_hr_fix_prompts(
    prompts: list[str],
    original_hr_prompt: str,
    original_prompt: str,
) -> list[str]:
    if original_prompt == original_hr_prompt:
        return list(prompts)
    return repeat_iterable_to_length([original_hr_prompt], len(prompts))
"""
`

If you update, it may overwrite this change and you may need to do it again. (I don't feel comfortable submitting a pull request as the Forge/A1111 code is way beyond my understanding and I don't really get why the original_prompt check exists.
